src/main/DynamicConfig.ts

```diff
@@ -299,13 +299,46 @@ export class DynamicConfig implements IDynamicConfig {
     }
 
     public async getRemoteValue<T>(key: string, type?: ObjectType): Promise<T> {
-        return this.source(key).then((source: ISource) => {
-            if (source.key !== undefined) {
-                return this.getValueFromResolver<T>(source.key, 'remote', type)
-            } else {
-                throw new errors.ResolverUnavailable(key)
-            }
-        })
+        const source = await this.source(key) // get source for key
+        const { key: sourceKey } = source
+        if (!sourceKey) {
+            throw new errors.ResolverUnavailable(key) // throw if key is undefined
+        }
+
+        const resolvedConfig = await this.getConfig() // get the current config (the cached version, that is)
+
+        const value = await this.getValueFromResolver<T>(
+            sourceKey,
+            'remote',
+            type,
+        ) // get new remote value for key
+
+        if (value === null) {
+            throw new errors.ResolverUnavailable(key) // if it doesn't exist, throw
+        }
+
+        const normalizedKey: string = Utils.normalizePath(key) // normalize/format key path
+
+        /*
+         build the normalized key path for the refreshed value to live at.
+         *note*: this is **required** to format the new value into a shape consumable by `setConfig()`.
+         */
+        const builtValue = ConfigBuilder.buildBaseConfigValue(source, value)
+
+        await this.setConfig(
+            ConfigUtils.setValueForKey(
+                normalizedKey,
+                builtValue,
+                resolvedConfig,
+                true,
+            ) as IRootConfigValue,
```

Comment:
Is there anyway to avoid casting to `IRootConfigValue` in this case / why is casting needed?

Reply:
Yep so the main reason right now is because of the function `setValueForKey`. [This function](https://github.com/creditkarma/dynamic-config/blob/8db80f2e3aa3a06c7bf4ae8831cffa5d73cc7480/src/main/utils/config.ts#L393) conditionally sets a key value, so it also conditionally returns a type. 
![image](https://user-images.githubusercontent.com/96585455/195696136-b4e7ba4f-c901-4760-ac64-0d663799647a.png)
We could separate this into 2 functions, or, merge the logic into 1, or do something else clever probably haha. 

But in the mean time, I chose casting because it would require the least amount of scope creep for this ticket. 

I don't love it either, though. 

I'd also love to update `setConfig` to return the newly updated config, then I could use that value in the `return` of the function to verify that the update was successful.

Reply:
That make sense, thanks for the explanation!